external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -2937,12 +2938,88 @@ fn module_call_impl(
     (call, return_)
 }
 
+/// If the constant that we are referencing has an `error` attribute, we need to change the type of
+/// the constant to a u64 since this will be compiled into a u64 error code.
+fn annotated_error_const(context: &mut Context, e: &mut T::Exp) {
```

Comment:
Taking another look at this I think we should rework this a bit:

- Move all feature gating to the `#[error]` attribute
- If it is a constant, check for `#[error]`. If it is `#[error]`, swap to a new `UnannotatedExp_::ErrorCode(Option((ModuleIdent, ConstantName))` Node. If None, it means just the loc
- Otherwise, join and check for the type to be `u64`. If it is not `u64` 
  - Custom help message
    - If it is a constant, suggest adding `#[error]`
    - If it is not, suggest binding to a  `#[error] const`
  - In either failure case, replace `e` with `UnresolvedError` to avoid double reporting  an invalid argument to `asset`

Reply:
Need to ponder on this a bit tonight, and lets chat about it tomorrow morning.

I'm not 100% sold that adding a new node to the AST here is buying us anything other than additional complexity (but maybe I'm missing something here? I probably am). If we did do this though I kinda would like to introduce this all the way up in naming when we convert `assert!(<cond>)` that way we can carry it all the way through.